File: src/lib/wow/db/caches/db-item-displays.ts
/**
 * Item display cache, ported from wow.export (src/js/db/caches/DBItemDisplays.js).
 */
import { write } from '@/lib/wow/log';

import { doOnce } from '../../formats/generics';
import { WDCReader } from '../wdc-reader';
import type { ModelDisplay } from './db-creatures';
import { getModelFileDataID } from './db-model-file-data';
import { ensureInitialized, getTextureFDIDsByMatID } from './db-texture-file-data';

const itemDisplays = new Map<number, ModelDisplay[]>();

/** Initialize item displays from ItemDisplayInfo.db2 */
export const initializeItemDisplays = doOnce('initializeItemDisplays', async () => {
  if (itemDisplays.size > 0) return;
  await ensureInitialized();

  write('Loading item textures...');
  const itemDisplayInfo = new WDCReader('DBFilesClient/ItemDisplayInfo.db2');
  await itemDisplayInfo.parse();

  // Using the texture mapping, map all model fileDataIDs to used textures.
  for (const [itemDisplayInfoID, itemDisplayInfoRow] of itemDisplayInfo.getAllRows()) {
    const modelResIDs = (itemDisplayInfoRow.ModelResourcesID as number[]).filter((e) => e > 0);
    if (modelResIDs.length === 0) continue;

    const matResIDs = (itemDisplayInfoRow.ModelMaterialResourcesID as number[]).filter((e) => e > 0);
    if (matResIDs.length === 0) continue;

    const modelFileDataIDs = getModelFileDataID(modelResIDs[0]);
    const textureFileDataIDs = getTextureFDIDsByMatID(matResIDs[0]);

    if (modelFileDataIDs !== undefined && textureFileDataIDs !== undefined) {
      for (const modelFileDataID of modelFileDataIDs) {
        const display: ModelDisplay = { ID: itemDisplayInfoID, textures: textureFileDataIDs };

        if (itemDisplays.has(modelFileDataID)) itemDisplays.get(modelFileDataID)!.push(display);
        else itemDisplays.set(modelFileDataID, [display]);
      }
    }
  }

  write('Loaded textures for %d items', itemDisplays.size);
});

/** Gets item skins from a given file data ID. */
export function getItemDisplaysByFileDataID(fileDataID: number): ModelDisplay[] | undefined {
  return itemDisplays.get(fileDataID);
}

export function getItemDisplayCacheStats(): { itemDisplays: number } {
  return { itemDisplays: itemDisplays.size };
}

export function resetItemDisplayCache(): void {
  itemDisplays.clear();
}

export default { initializeItemDisplays, getItemDisplaysByFileDataID };

File: golang/internal/wow/db/caches/db_item_displays.go
package caches

import (
	"context"
	"sync"

	"github.com/pqhuy98/wow-converter/internal/wow/db"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

var (
	itemDisplays   = make(map[uint32][]ModelDisplay)
	itemOnce       sync.Once
	itemErr        error
)

// InitializeItemDisplays loads ItemDisplayInfo.db2.
func InitializeItemDisplays(ctx context.Context) error {
	itemOnce.Do(func() {
		if len(itemDisplays) > 0 {
			return
		}
		if err := InitializeModelFileData(ctx); err != nil {
			itemErr = err
			return
		}
		if err := EnsureTextureInitialized(ctx); err != nil {
			itemErr = err
			return
		}

		log.Write("Loading item textures...")
		reader := db.NewWDCReader("DBFilesClient/ItemDisplayInfo.db2", nil)
		if err := reader.Parse(ctx, nil); err != nil {
			itemErr = err
			return
		}

		for itemDisplayInfoID, row := range reader.GetAllRows() {
			modelResIDs := filterPositive(toUint32Slice(row["ModelResourcesID"]))
			if len(modelResIDs) == 0 {
				continue
			}
			matResIDs := filterPositive(toUint32Slice(row["ModelMaterialResourcesID"]))
			if len(matResIDs) == 0 {
				continue
			}
			modelFileDataIDs := GetModelFileDataID(modelResIDs[0])
			textureFileDataIDs := GetTextureFDIDsByMatID(matResIDs[0])
			if modelFileDataIDs == nil || textureFileDataIDs == nil {
				continue
			}
			for _, modelFileDataID := range modelFileDataIDs {
				display := ModelDisplay{ID: itemDisplayInfoID, Textures: textureFileDataIDs}
				itemDisplays[modelFileDataID] = append(itemDisplays[modelFileDataID], display)
			}
		}
		log.Write("Loaded textures for %d items", len(itemDisplays))
	})
	return itemErr
}

// GetItemDisplaysByFileDataID returns item skins for a file data ID.
func GetItemDisplaysByFileDataID(fileDataID uint32) []ModelDisplay {
	return itemDisplays[fileDataID]
}

// GetItemDisplayCacheStats returns cache stats.
func GetItemDisplayCacheStats() int {
	return len(itemDisplays)
}

// ResetItemDisplayCache clears item display cache.
func ResetItemDisplayCache() {
	itemDisplays = make(map[uint32][]ModelDisplay)
	itemOnce = sync.Once{}
	itemErr = nil
}
